File: C#_Everything/DS Modules/LinkList.cs
namespace My_DS_Modules;

class LinkList<T> where T : notnull
{
  /// <summary>
  ///   Node data structure
  /// </summary>
  /// <param name="value">Node value</param>
  public class Node(T value)
  {
    public T Value { get; set; } = value;
    public Node? Prev { get; set; }
    public Node? Next { get; set; }
  };

  public Node? Head { get; private set; } = null;
  public Node? Tail { get; private set; } = null;
  public int Count { get; private set; } = 0;

  /// <summary>
  ///   Add node before Head node
  /// </summary>
  /// <param name="Value">Node value</param>
  public void AddFirst(T Value)
  {
    // Create new node
    Node node = new(Value);

    if (Head is null) // checks for empty link list
    {
      Head = Tail = node;
      Count = 1;
      return;
    }

    /* Senerio: Head->▭<-Tail */
    /* New node: ▭ ▭ */
    /* Head.Prev to Node: ▭-Head->▭<-Tail */
    /* Node.Next to Head: ▭=Head->▭<-Tail */
    /* Head as node: Head->▭=▭<-Tail */
    AddReLink(asBefore: node, asAfter: Head);
    Head = node;
    Count++;
  }

  /// <summary>
  ///   Add node after Tail node
  /// </summary>
  /// <param name="Value">Node value</param>
  public void AddLast(T Value)
  {
    // Create new node
    Node node = new(Value);

    if (Tail is null) // checks for empty link list
    {
      Head = Tail = node;
      Count = 1;
      return;
    }

    /* Senerio: Head->▭<-Tail */
    /* New node: ▭ ▭ */
    /* Node.Prev to Tail: Head->▭<-Tail-▭ */
    /* Tail.Next to Node: Head->▭<-Tail=▭ */
    /* Tail as node: Head->▭=▭<-Tail */
    AddReLink(asBefore: Tail, asAfter: node);
    Tail = node;
    Count++;
  }

  /// <summary>
  ///   Remove Head node
  /// </summary>
  public void RemoveFirst()
  {
    if (Head is null) return; // checks for empty link list
    if (Head.Next is null) // Only one node
    {
      Clear();
      return;
    }

    /* Senerio: Head->▭=▭<-Tail */
    /* Head to Head.Next: ▭=Head->▭<-Tail*/
    /* Head.Prev to null: Head->▭<-Tail*/
    Head = Head.Next;
    Head.Prev = null;
    Count--;
  }

  /// <summary>
  ///   Remove tail node
  /// </summary>
  public void RemoveLast()
  {
    if (Tail is null) return; // checks for empty link list
    if (Tail.Prev is null) // Only one node
    {
      Clear();
      return;
    }

    /* Senerio: Head->▭=▭<-Tail */
    /* Tail to Tail.Prev: Head->▭<-Tail=▭*/
    /* Tail.Next to null: Head->▭<-Tail*/
    Tail = Tail.Prev;
    Tail.Next = null;
    Count--;
  }

  /// <summary>
  ///   Insert new node before ref node.
  /// </summary>
  /// <param name="node">Add from</param>
  /// <param name="value">New node value</param>
  public void InsertBefore(Node? node, T value)
  {
    /* Only one or no Prev node (is Head) equivalent to Add first*/
    if (node is null || node.Prev is null)
    {
      AddFirst(value);
      return;
    }

    // Create new node
    Node newNode = new(value);

    // Head->▭=▭=▭=▭...=▭<-Tail
    //            ▭       
    // link new node first to before and after node: so we still has ref
    LinkNewNode(newNode, node);
    // Head->▭=▭=▭=▭...=▭<-Tail
    //          |  |
    //          --▭
    // Relink newNode Prev node and Next Node to new node
    node.Prev.Next = newNode;
    node.Prev = newNode;
    // Head->▭=▭=▭=▭=▭...=▭<-Tail
    //             |INSERTED
    Count++;
  }

  /// <summary>
  ///   Insert new node before ref node.
  /// </summary>
  /// <param name="node">Add from</param>
  /// <param name="value">New node value</param>
  public void InsertAfter(Node? node, T value)
  {
    /* Only one or no Next node (is Tail) equivalent to Add last*/
    if (node is null || node.Next is null)
    {
      AddLast(value);
      return;
    }

    // Create new node
    Node newNode = new(value);

    // Head->▭=▭=▭=▭...=▭<-Tail
    //            ▭       
    // link new node first to before and after node: so we still has ref
    LinkNewNode(newNode, node.Next);
    node.Next.Prev = newNode;
    node.Next = newNode;
    Count++;
  }

  private static void AddReLink(Node asBefore, Node asAfter)
  {
    asAfter.Prev = asBefore;
    asBefore.Next = asAfter;
  }

  private static void LinkNewNode(Node newNode, Node newNodeAfter)
  {
    newNode.Prev = newNodeAfter.Prev;
    newNode.Next = newNodeAfter;
  }

  public IEnumerable<T> Forward()
  {
    for (var current = Head; current is not null; current = current.Next)
      yield return current.Value;
  }

  public IEnumerable<T> Backward()
  {
    for (var current = Tail; current is not null; current = current.Prev)
      yield return current.Value;
  }

  public Node? Find(T value)
  {
    for (var current = Head; current is not null; current = current.Next)
      if (Equals(current.Value, value)) return current; // Match, immediate return

    return null; // No match or start empty
  }

  public bool Contains(T value)
  => Find(value) is not null;

  public void Remove(Node node)
  {
    if (node == Head) // Head same as Remove first
    {
      RemoveFirst();
      return;
    }

    if (node == Tail) // Tail same as Remove first
    {
      RemoveLast();
      return;
    }

    // Prev node next to Next node
    // Next node Prev to Prev node
    var prev = node.Prev;
    var next = node.Next;

    if (prev is not null) prev.Next = next;
    if (next is not null) next.Prev = prev;

    // node.Prev, node.Next is still link to node.Prev.Next & node.Next.Prev
    // Garbage collect
    node.Next = null;
    node.Prev = null;

    Count--;
  }

  /// <summary>
  ///   Remove node by value that is the first found
  /// </summary>
  /// <param name="value">Target value</param>
  /// <returns>bool: success</returns>
  public bool Remove(T value)
  {
    // Find target node
    Node? node = Find(value);

    // No target to remove
    if (node is null) return false; // Unsuccess

    // Has target Remove it
    Remove(node);

    // Success
    return true;
  }

  public void Clear()
  {
    Head = Tail = null;
    Count = 0;
  }
}

File: Python_Everything/Ds_Module/LinkList.py
from typing import Generic, List, Optional, TypeVar

T = TypeVar('T');

class Node(Generic[T]):
  def __init__(
    self, Data: T, 
    Prev: Optional[Node[T]] = None, 
    Next: Optional[Node[T]] = None
  ):
    self.Data = Data
    self.Prev = Prev
    self.Next = Next


class LinkList(Generic[T]):

  # Field 
  def __init__(self):
    self.Head: Optional[Node[T]] = None;
    self.Tail: Optional[Node[T]] = None;
    self.Count: int = 0;


  def AddFirst(self, data: T) -> None:
    # New node
    node = Node(data)

    # No node
    if self.Head is None:
      self.Head = self.Tail = node
      Count = 1
      return
    
    # Relink together
    self.__AddRelink(asBefore=node, asAfter=self.Head)

    # New Head
    self.Head = node
    self.Count += 1


  def AddLast(self, data: T) -> None:
    # New node
    node = Node(data)

    # No node
    if(self.Tail is None):
      self.Head = self.Tail = node
      Count = 1
      return
    
    # Relink together
    self.__AddRelink(asBefore=self.Tail, asAfter=node)

    # New tail
    self.Tail = node
    self.Count += 1
    

  def __AddRelink(self, asBefore: Node[T], asAfter: Node[T]) -> None: # Not Optional
    asAfter.Prev = asBefore
    asBefore.Next = asAfter


  def InsertBefore(self, refNode: Node[T], data: T) -> None:
    if refNode is None: 
      raise ValueError("Null Reference error: refNode must be none empty")
      return

    # is head or one Node
    if refNode.Prev is None:
      self.AddFirst(data)
      return
    
    # defualt: Two Node not link
    newNode = Node(data);
    newNode.Prev = refNode.Prev
    newNode.Next = refNode
    
    self.InsertNodeLink(asNewNode=newNode, asNewNodeAfter=refNode)

    self.Count += 1


  def InsertAfter(self, refNode: Node[T], data: T) -> None:
    if refNode is None: 
      raise ValueError("Null Reference error: refNode must be none empty")
      return

    # is head or one Node
    if refNode.Prev is None:
      self.AddFirst(data)
      return
    
    # defualt: Two Node not link
    newNode = Node(data);
    newNode.Prev = refNode
    newNode.Next = refNode.Next

    self.InsertNodeLink(asNewNode=newNode, asNewNodeAfter=refNode.Next)
    self.Count += 1;


  def InsertNodeLink(self, asNewNode: Node[T], asNewNodeAfter: Node[T]):
    asNewNodeAfter.Prev.Next = asNewNode
    asNewNodeAfter.Prev = asNewNode

  
  def RemoveFirst(self) -> None:
    if self.Head is None:
      return
    
    # One node
    if self.Head.Next is None:
      self.Clear()
      return

    self.Head = self.Head.Next
    self.Head.Prev = None
    self.Count -= 1;


  def RemoveLast(self) -> None:
    if self.Tail is None:
      return
    
    # One node
    if self.Tail.Prev is None:
      self.Clear()
      return

    self.Tail = self.Tail.Prev
    self.Tail.Next = None
    self.Count -= 1;


  def Remove(self, refNode:Node[T]) -> None:
    if refNode is None: 
      raise ValueError("Null Reference error: refNode must be none empty")
      return

    # is Head
    if refNode.Prev is None:
      self.RemoveFirst()
      return
    
    # is tail
    if refNode.Next is None:
      self.RemoveLast()
      return
    
    refNode.Next.Prev = refNode.Prev
    refNode.Prev.Next = refNode.Next

    self.Count -= 1
    

  def Contains(self, data: T) -> bool:
    current = self.Head;
    while current is not None:
      # Match
      if current.Data == data:
        return True
      
      # Check Next
      current = current.Next
    return False
  

  def Find(self, data: T) -> Optional[Node[T]]:
    current = self.Head;
    while current is not None:
      # Match
      if current.Data == data:
        return current
      
      # Check Next
      current = current.Next
    return None
  

  def Forward(self) -> List[T]:
    current = self.Head
    returnList:List[T] = []
    while current is not None:
      returnList.append(current.Data)
      
      current = current.Next
    return returnList
  

  def Backward(self) -> List[T]:
    current = self.Tail
    returnList:List[T] = []
    while current is not None:
      returnList.append(current.Data)
      
      current = current.Prev
    return returnList


  def GetCount(self) -> int:
    return self.Count


  def Clear(self) -> None:
    self.Head = self.Tail = None
    self.Count = 0